content/blog/2022-04-04-develop-and-test-data-analytics-pipelines-on-localstack/index.md

```diff
@@ -0,0 +1,236 @@
+
+## Deploying Locally
+With our data pipeline stack defined and external resource mocks in order, it's time to deploy everything locally using `cdklocal`.
+
+### Define a Local App
+To keep things tidy, we'll create a [separate stack](https://github.com/localstack/support-bot-analytics/blob/simplified-blog/deployments/cdk/external_test_resources_stack.py) for the test logger Lambda and deploy it alongside the data pipeline stack under a single CDK app specifically for local testing. This way the data pipeline stack itself stays identical to what we deploy to AWS in production.
+
+```python
+# local_app.py
+
+def main():
+    # configuration variables for connecting to local mock Tinybird server
+    port = 5111
+    host = os.getenv("TINYBIRD_MOCK_HOST") or "host.docker.internal"
+    tinybird_mock_path = "tinybird"
+    tinybird_mock_url = f"http://{host}:{port}/{tinybird_mock_path}"
+    
+    app = App()
+
+    external_resources = ExternalTestResourcesStack(
+        app,
+        "ExternalTestResourcesStack",
+        env=constants.ENV_LOCAL,
+    )
+
+    data_pipeline = DataPipelineStack(
+        app,
+        "DataPipelineStack",
+        env=constants.ENV_LOCAL,
+        monitored_log_group_arn=Fn.import_value("test-logger-log-group-arn"),
+        tinybird_auth_token="dummy value",
+        tinybird_url=tinybird_mock_url,
+    )
+    data_pipeline.add_dependency(external_resources)
+
+    app.synth()
+
+
+if __name__ == "__main__":
+    main()
+```
+
+Note how the `DataPipelineStack` uses the log group ARN value imported from the `ExternalTestResources` stack.
+
+### Deploy Locally
+Alright, let's deploy! First, we need to start LocalStack, bootstrap the local CDK environment, and deploy the logger Lambda. Please note that you will need [LocalStack Pro](https://localstack.cloud/pricing/) enabled for the log subscription feature to work.
+
+```bash
+LOCALSTACK_API_KEY="your key here" localstack start -d
+cdklocal bootstrap --app "python local_app.py"
+cdklocal deploy --app "python local_app.py" ExternalTestResourcesStack
+```
+You should see output similar to the following:
+
+```
+ ✅  ExternalTestResourcesStack
+
+✨  Deployment time: 6.78s
+
+Outputs:
+ExternalTestResourcesStack.testloggerlambdaname = ExternalTestResourcesStack-testlogger39117BFF-c39d4d5a
+ExternalTestResourcesStack.testloggerloggrouparn = arn:aws:logs:us-east-2:000000000000:log-group:/aws/lambda/ExternalTestResourcesStack-testlogger39117BFF-c39d4d5a:*
+Stack ARN:
+arn:aws:cloudformation:us-east-2:000000000000:stack/ExternalTestResourcesStack/b5143830
+
+✨  Total time: 13.53s
+```
+Before we can deploy the data pipeline itself we have to invoke the logger Lambda. Cloudwatch log groups for Lambda functions aren't created until the function is executed for the first time. Unless we do this step first the deployment will fail because CDK won't be able to locate the log group resource.
```

Comment:
```suggestion
Before we can deploy the data pipeline itself we have to invoke the logger Lambda. CloudWatch log groups for Lambda functions aren't created until the function is executed for the first time. Unless we do this step first the deployment will fail because CDK won't be able to locate the log group resource.
```